Add `forward`

badeend opened the pull request:
During the plumber's summit I noticed that a forwarding mechanism was up for consideration in preview2.1.
So, here's my take.

badeend committed:
Add `forward`

pchickey commented on wit/streams.wit:
```diff
@@ -259,4 +259,40 @@ interface streams {
+    /// yet it does not change the ownership structure of them. I.e. they remain
+    /// children of their parent resources. If the parents place any lifetimes
+    /// restrictions on their children, those continue to apply.
+    forward: func(src: input-stream, dst: output-stream) -> future-forward-result;
```
These arguments will pass ownership of both the input-stream and output-stream. I believe sending an owned input-stream is correct here. In some use cases, we need a way to permit further use of the output-stream after forwarding has either finished or failed. We can do that by either making this parameter a borrow of the output-stream, or we get an owned output-stream in the resolution of future-forward-result.get

badeend replied:
Changing `dst: output-stream` from an owned handle to a borrowed handle will introduce multithreaded access to the stream. I can't see a way why one would want this as there's no way coordinate between the writes from different tasks/threads. From end-user and implementation complexity perspective I'd go for: consume the stream and hand it back later. Or even simpler: document & enforce that while the forward is in progress any attempt to access the borrowed streams will trap.

That being said, the way I designed it, the host is guaranteed that no one else can touch the streams ever again. In my mind this is a "feature", as the host can then permanently fuse them together. This may enable additional optimizations not possible otherwise.

An example that triggered me to file this PR in the first place is: kTLS. Piping a TLS stream into a Socket stream could "automagically" offload TLS to the kernel (or even the NIC). Requiring the output-stream to remain usable after the input-stream has ended would prevent this optimization. To be clear, I haven't actually implemented this, so assign as much value to this use-case as you want.

Anyway, I can see merit in both designs. Maybe they can coexist?

pchickey replied:
I hadn't thought about that case, but I have heard of some use cases for wanting to use the destination stream after  the source stream is consumed. So, I guess they can both coexist. Are you up for introducing both variants in this PR?

pchickey commented on wit/streams.wit:
```diff
@@ -259,4 +259,40 @@ interface streams {
+        ///   has already been taken before.
+        /// - `some(ok(value))`: when the future is ready with a value. This is
+        ///   returned only once.
+        get: func() -> option<result<stream-error>>;
```
We also need this get function to have a way to represent "youve already retrieved the result of this and i cant give it to you again", in http we used an outer result for that, so it would end up like `option<result<result<stream-error>, _>>`

badeend replied:
Indeed, that's documented as:

```
- `some(error(_))`: when the future is resolved, but the result value has already been taken before.
```

pchickey replied:
oh, sorry, I read the type signature before the docs here. How do we represent the case where the forward completed with success? there is no stream-error to represent success, you need a `result<stream-error>` for that.

badeend replied:
Well "success" in this case is when one of the streams ended with `closed`. Which is already captured in the `stream-error` type.

pchickey replied:
OK, but as a user I would want to distinguish between whether the input-stream closed (expected outcome of the operation) and the output-stream closed (unexpected). When the expected outcome happens, we should return an `ok`.

Additionally I'd consider returning a `variant forward-error { source(stream-error), destination(stream-error) }` to distinguish where the error came from - we shouldnt throw away information that the user will need to debug.

badeend replied:
> OK, but as a user I would want to distinguish between whether the input-stream closed (expected outcome of the operation) and the output-stream closed (unexpected). When the expected outcome happens, we should return an ok.

Fair enough.

> Additionally I'd consider returning a variant forward-error { source(stream-error), destination(stream-error) } to distinguish where the error came from - we shouldnt throw away information that the user will need to debug.

I took this design from `splice` which already tosses errors from both streams on the same pile. Are you sure we want to introduce this inconsistency & complexity in return types? Or is it OK that say that whenever a consumer really cares about which stream failed, they can just attempt `input-stream::read(0)` or `output-stream::check-write()` after the fact?

pchickey reviewed (commented):
Looks like a great start! Thanks for writing this up.

sunfishcode commented on wit/streams.wit:
```diff
@@ -259,4 +259,40 @@ interface streams {
+    /// restrictions on their children, those continue to apply.
+    forward: func(src: input-stream, dst: output-stream) -> future-forward-result;
+
+    /// Represents a future which may eventually return trailers, or an error.
```
Since this is wasi-io and just focused on streams, it doesn't know anything about trailers per se.

If you want to mention how this interacts with trailers in wasi-http, I suggest putting that in a separate paragraph and introducing it with "In wasi-http, ...".

badeend replied:
Hah, no I don't want to mention HTTP here.
I copy-pasted the future type from wasi-http and apparently I forgot to edit this comment. :)

badeend commented:
I've updated the definitions. Pretty much the entire text has changed so please check again (sorry).

Noteworty changes:
- There are now two dinstinct functions.
- The future now returns an additional `result`.
- I added a `flush-on-block` parameter. TBH, I'm not quite happy with this as-is, because there are a bazillion flushing strategies. The current version uses I/O readiness for its heuristic. Ideally, the input stream should have a native way to to have a way to return a flush signal from the `read` operation.
- `forward` states: "Any attempt to access or drop the streams in the meantime will trap.". This is the simplest to implement. Another valid strategy could be to say that while the forward is in progress, any read/write attempt will return their EWOULDBLOCK equivalent until the forward is finished. This latter solution is less surprising to end users, but more difficult to implement.
- Dropping the future may trap depending on its state.
- The future can be canceled.

I have _not_ added the `forward-error` variant (yet), see above.

badeend committed:
Remove flush-on-block

badeend commented:
After discussion in #73, I realized my interpretation of flush was not quite correct. I've removed the `flush-on-block` flag.

ricochet commented:
Thank you for contributing to WASI! 

As part of our proposal management process, we decided to consolidate all phase 3 proposals into a single monorepo at [WebAssembly/wasi](https://github.com/WebAssembly/wasi) to better manage the proposals that have reached phase 3.

We have completed migrating all phase 3 proposals to [WebAssembly/wasi](https://github.com/WebAssembly/wasi) including their issues, but PRs need to be re-opened in the new repository. This is a GitHub limitation and we're sorry for the inconvenience. Please re-open this PR on [WebAssembly/wasi](https://github.com/WebAssembly/wasi). Thank you!
